docs/guides/routing.rst

```diff
@@ -1,4 +1,121 @@
 Routing
 #######
 
-Add new pages and customize URLs.
\ No newline at end of file
+Elgg has two mechanisms to respond to HTTP requests that don't already go through the
+:doc:`/design/actions` and :doc:`/guides/views/simplecache` systems.
+
+URL Identifier and Segments
+===========================
+
+After removing the site URL, Elgg splits the URL path by ``/`` into an array. The first
+element, the **identifier**, is shifted off, and the remaining elements are called the
+**segments**. For example, if the site URL is ``http://example.com/elgg/``, the URL
+``http://example.com/elgg/blog/owner/jane?foo=123`` produces:
+
+Identifier: ``'blog'``. Segments: ``['owner', 'jane']``. (the query string parameters are
+available via ``get_input()``)
+
+The site URL (home page) is a special case that produces an empty string identifier and
+an empty segments array.
+
+
+Page Handler
+============
+
+To handle all URLs that begin with a particular identifier, you can register a function to
+act as a :doc:`/guides/pagehandler`. When the handler is called, the segments array is
+passed in as the first argument.
+
+The following code registers a page handler for "blog" URLs and shows how one might route
+the request to a resource view.
+
+.. code:: php
+
+   elgg_register_page_handler('blog', 'blog_page_handler');
+
+   function blog_page_handler(array $segments) {
+        // if the URL is http://example.com/elgg/blog/view/123/my-blog-post
+        // $segments contains: ['view', '123', 'my-blog-post']
+
+        $subpage = elgg_extract(0, $segments);
+        if ($subpage === 'view') {
+
+            // use a view for the page logic to allow other plugins to easily change it
+            set_input('guid', (int)elgg_extract(1, $segments));
+            echo elgg_view('resources/blog/view');
+
+            // in page handlers, return true says, "we've handled this request"
+            return true;
+        }
+
+        // ... handle other subpages
+   }
+
+
+The ``route`` Plugin Hook
+=========================
+
+The ``route`` plugin hook is triggered earlier, before page handlers are called. The URL
+identifier is given as the type of the hook. This hook can be used to modify the identifier
+or segments, to take over page rendering completely, or just to add some logic before the
+request is handled elsewhere.
+
+Generally devs should use a page handler unless they need to affect a single page or a wider variety of URLs.
```

Comment:
I don't think we need to make this recommendation. People can use whichever works best for them.